### PR TITLE
IGNITE-16691 Storage Configuration refactoring and improvements

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-16691

### CODE REVIEW

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java [hunk 1]
```diff
@@ -36,45 +42,94 @@
...
+    public TableStorage createTable(TableConfiguration tableCfg) throws StorageException {
+        TableView tableView = tableCfg.value();
+
+        assert tableView.dataStorage().name().equals(ENGINE_NAME) : tableView.dataStorage().name();
+
+        RocksDbDataStorageView dataStorageView = (RocksDbDataStorageView) tableView.dataStorage();
+
+        RocksDbDataRegion dataRegion = regions.get(dataStorageView.dataRegion());
+
+        if (dataRegion == null) {
```
Comment (ibessonov) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java:121]: Technically speaking, this should be impossible and hence should be replaced with assertion. What do you think?

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineFactory.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+/**
+ * Implementation for creating {@link RocksDbStorageEngine}'s.
+ */
+public class RocksDbStorageEngineFactory implements StorageEngineFactory {
+    /** {@inheritDoc} */
+    @Override
+    public StorageEngine createEngine(ConfigurationRegistry configRegistry, Path storagePath) throws StorageException {
+        RocksDbStorageEngineConfiguration engineConfig = configRegistry.getConfiguration(RocksDbStorageEngineConfiguration.KEY);
+
+        if (engineConfig == null) {
```
Comment (ibessonov) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineFactory.java:36]: How's this possible? Should be assertion as well

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/configuration/schema/RocksDbDataStorageConfigurationSchema.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import org.apache.ignite.configuration.validation.Immutable;
+import org.apache.ignite.internal.storage.rocksdb.RocksDbStorageEngine;
+
+/**
+ * Data storage configuration for {@link RocksDbStorageEngine}.
+ */
+@PolymorphicConfigInstance(ENGINE_NAME)
+public class RocksDbDataStorageConfigurationSchema extends DataStorageConfigurationSchema {
+    /** Data region. */
+    @Immutable
```
Comment (ibessonov) [inline modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/configuration/schema/RocksDbDataStorageConfigurationSchema.java:35]: Should this really be immutable? Old implementation allowed changing regions...

Comment (ibessonov) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/configuration/schema/RocksDbDataStorageConfigurationSchema.java:35]: And..., maybe we will remove regions from RocksDb storages one day and it's all irrelevant

File: modules/api/src/main/java/org/apache/ignite/configuration/schemas/store/DataStorageConfigurationSchema.java [hunk 4]
```diff
@@ -17,28 +17,18 @@
...
- * Root configuration for data storages.
+ * Configuration schema for data storage.
  */
-@ConfigurationRoot(rootName = "db", type = ConfigurationType.DISTRIBUTED)
+@PolymorphicConfig
 public class DataStorageConfigurationSchema {
-    /** Name of the default data region. */
-    public static final String DEFAULT_DATA_REGION_NAME = "default";
+    /** Default data storage name. */
+    public static final String DEFAULT_DATA_STORAGE_NAME = "rocksdb";
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/api/src/main/java/org/apache/ignite/configuration/schemas/store/DataStorageConfigurationSchema.java:29]: Ok.

File: modules/page-memory/src/integrationTest/java/org/apache/ignite/internal/pagememory/persistence/ItBplusTreePageMemoryImplTest.java [hunk 5]
```diff
@@ -54,7 +49,7 @@ protected PageMemory createPageMemory() throws Exception {
 
         return new PageMemoryImpl(
                 new UnsafeMemoryProvider(null),
-                (PageMemoryDataRegionConfiguration) fixConfiguration(dataRegionCfg),
+                fixConfiguration(dataRegionCfg),
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/page-memory/src/integrationTest/java/org/apache/ignite/internal/pagememory/persistence/ItBplusTreePageMemoryImplTest.java:52]: You're right, I'll take it out.

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/MockedStructuresTest.java [hunk 6]
```diff
@@ -178,6 +194,16 @@ void before() throws NodeStoppingException {
             });
         };
 
+        dsm = new DataStorageManager(configRegistry, workDir) {
+            /** {@inheritDoc} */
+            @Override
+            protected Iterable<StorageEngineFactory> engineFactories() {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/exec/MockedStructuresTest.java:200]: Discussed in person, decided to leave as is.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    ) {
+        this.clusterConfigRegistry = clusterConfigRegistry;
+        this.storagePath = storagePath;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void start() {
+        for (StorageEngineFactory engineFactory : engineFactories()) {
+            StorageEngine engine = engineFactory.createEngine(clusterConfigRegistry, storagePath);
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java:62]: Fix it.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+    protected Iterable<StorageEngineFactory> engineFactories() {
+        return ServiceLoader.load(StorageEngineFactory.class);
+    }
+
+    /**
+     * Returns the data storage engine by name.
+     *
+     * @param name Engine name.
+     */
+    public @Nullable StorageEngine engine(String name) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java:98]: I'll take it away.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/StorageEngine.java [hunk 9]
```diff
@@ -27,32 +26,31 @@
  */
 public interface StorageEngine {
     /**
-     * Starts the engine.
+     * Returns the unique name of the storage engine.
+     *
+     * <p>NOTE: Used to map {@link DataStorageConfigurationSchema#name} to {@link StorageEngine#name}.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/engine/StorageEngine.java:31]: Fix it.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PageMemoryDataRegion.java [hunk 10]
```diff
@@ -48,7 +47,7 @@ public PageMemoryDataRegion(PageMemoryDataRegionConfiguration cfg, PageIoRegistr
 
     /** {@inheritDoc} */
     @Override
-    public void stop() {
+    public void stop() throws Exception {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PageMemoryDataRegion.java:50]: No, fix it.

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/configuration/schema/PageMemoryStorageEngineConfigurationSchema.java [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+import org.apache.ignite.configuration.annotation.Name;
+import org.apache.ignite.configuration.annotation.NamedConfigValue;
+import org.apache.ignite.configuration.validation.ExceptKeys;
+import org.apache.ignite.internal.pagememory.configuration.schema.PageMemoryDataRegionConfigurationSchema;
+import org.apache.ignite.internal.storage.pagememory.PageMemoryStorageEngine;
+
+/**
+ * Root configuration for {@link PageMemoryStorageEngine}.
+ */
+@ConfigurationRoot(rootName = ENGINE_NAME, type = DISTRIBUTED)
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/configuration/schema/PageMemoryStorageEngineConfigurationSchema.java:34]: You're not necessarily right, I've corrected that.

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java (hunk 1 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngine.java:121]: Replaced with assertion

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/configuration/schema/RocksDbDataStorageConfigurationSchema.java (hunk 3 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/configuration/schema/RocksDbDataStorageConfigurationSchema.java:35]: You're right, it's not needed, removed it.

File: modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineFactory.java (hunk 2 above)
Comment (tkalkirill) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineFactory.java]: Replaced with: **@PageMemoryDataRegionValidator** and **@RocksDbDataRegionValidator**

Comment (ibessonov) [reply modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbStorageEngineFactory.java]: I should refactor our approach to validators, too many weird annotations =)

File: modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PageMemoryStorageEngineFactory.java [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+ */
+public class PageMemoryStorageEngineFactory implements StorageEngineFactory {
+    /** {@inheritDoc} */
+    @Override
+    public StorageEngine createEngine(ConfigurationRegistry configRegistry, Path storagePath) throws StorageException {
+        PageMemoryStorageEngineConfiguration engineConfig = configRegistry.getConfiguration(PageMemoryStorageEngineConfiguration.KEY);
+
+        assert engineConfig != null;
+
+        PageIoRegistry ioRegistry = new PageIoRegistry();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/PageMemoryStorageEngineFactory.java:39]: Will do in the future, yes.

File: modules/storage-api/README.md [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+    * Implement `org.apache.ignite.internal.configuration.ConfigurationModule`;
+* Add services (which are loaded via `java.util.ServiceLoader.load(java.lang.Class<S>)`):
+    * Implementation of `org.apache.ignite.internal.storage.engine.StorageEngineFactory`;
+    * Implementation of `org.apache.ignite.internal.configuration.ConfigurationModule`.
+
+Take `org.apache.ignite.internal.storage.chm.TestConcurrentHashMapStorageEngine` as an example.
+
+## Usage
+
+Used in table configuration see `org.apache.ignite.configuration.schemas.table.TableConfigurationSchema.dataStorage`.
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/README.md:29]: Discussed personally what should be replaced.

File: modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/TableValidatorImpl.java [hunk 14]
```diff
@@ -63,70 +58,26 @@ public void validate(TableValidator annotation, ValidationContext<NamedListView<
...
-     * @param dbCfg      Data storage configuration.
-     * @param regionName Data region name.
-     * @return Data region configuration.
-     */
-    private static DataRegionView dataRegion(DataStorageView dbCfg, String regionName) {
-        if (regionName.equals(DEFAULT_DATA_REGION_NAME)) {
-            return dbCfg.defaultRegion();
+    private static void validateDataStorage(@Nullable TableView oldTable, TableView newTable, ValidationContext<?> ctx) {
+        if (oldTable != null) {
+            String oldDataStorage = oldTable.dataStorage().name();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/schema/src/main/java/org/apache/ignite/internal/schema/configuration/TableValidatorImpl.java]: Yes, I did.

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/validation/ValidationUtil.java [hunk 15]
```diff
@@ -99,12 +101,14 @@ private void validate(InnerNode lastInnerNode, String fieldName, Object val) {
```
Comment (ibessonov) [inline modules/configuration/src/main/java/org/apache/ignite/internal/configuration/validation/ValidationUtil.java:101]: Please use schemaType in the code above

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java [hunk 16]
```diff
@@ -315,7 +326,8 @@
     private final Map<Class<?>, SchemaClassesInfo> schemasInfo = new HashMap<>();
 
     /** Class generator instance. */
-    private final ClassGenerator generator = ClassGenerator.classGenerator(getClass().getClassLoader());
+    private final ClassGenerator generator = ClassGenerator.classGenerator(getClass().getClassLoader())
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/asm/ConfigurationAsmGenerator.java]: Done

File: modules/configuration/src/main/java/org/apache/ignite/internal/configuration/validation/ValidationUtil.java (hunk 15 above)
Comment (tkalkirill) [reply modules/configuration/src/main/java/org/apache/ignite/internal/configuration/validation/ValidationUtil.java:101]: Done

File: modules/configuration/src/test/java/org/apache/ignite/internal/configuration/validation/ValidationUtilTest.java [hunk 17]
```diff
@@ -127,73 +172,162 @@ public void validateLeafNode() {
...
+        }
+
+        static int compareByCurrentKey(ValidationIssue o1, ValidationIssue o2) {
+            ExValidationIssue ex1 = (ExValidationIssue) o1;
+            ExValidationIssue ex2 = (ExValidationIssue) o2;
+
+            return ex1.currentKey.compareTo(ex2.currentKey);
+        }
+
+        static List<ExValidationIssue> compareByCurrentKey(List<ValidationIssue> issues) {
```
Comment (ibessonov) [inline modules/configuration/src/test/java/org/apache/ignite/internal/configuration/validation/ValidationUtilTest.java:326]: Should it be "sortedByCurrentKey"?

**Review by ibessonov — APPROVED** [review]

**Pull request merged**